### PR TITLE
feat(docker): provide a Debian-based slim image

### PR DESCRIPTION
**Pull request by nejch**
Closes https://github.com/python-gitlab/python-gitlab/issues/1852.

Tested in https://gitlab.com/nejc/python-gitlab/-/pipelines/453883269 with a fake tag, resulting images:

https://gitlab.com/nejc/python-gitlab/container_registry/2676067

@rassie hopefully this covers your use case.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -53,8 +53,23 @@ From GitLab:
...
-----------------------
+Using the docker images
+-----------------------
+
+``python-gitlab`` provides Docker images in two flavors, based on the ``-alpine`` and ``-slim``
+python `base images <https://hub.docker.com/_/python>`__. The default tag is alpine, but you
+can explicitly use the alias (see below).
+
+The alpine is image is smaller, but you may want to use the Debian-based ``-slim`` tag if you
+are running into issues or need a more complete environment with a bash shell, such as in CI.
```
Comment (JohnVillalovos) [inline README.rst:64]:
```suggestion
The alpine image is smaller, but you may want to use the Debian-based ``-slim`` tag if you
are running into issues or need a more complete environment with a bash shell, such as in a CI.
```

**Review by JohnVillalovos — CHANGES_REQUESTED** [review]

**Review by JohnVillalovos — APPROVED** [review]

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -2,26 +2,30 @@ image: python:3.10
...
   script:
-    - mkdir /root/.docker && echo "{\"auths\":{\"$CI_REGISTRY\":{\"username\":\"$CI_REGISTRY_USER\",\"password\":\"$CI_REGISTRY_PASSWORD\"}}}" > /root/.docker/config.json
-    - /ko-app/crane cp $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG $CI_REGISTRY_IMAGE:latest
-  only:
-    - tags
+    - crane auth login -u $CI_REGISTRY_USER -p $CI_REGISTRY_PASSWORD $CI_REGISTRY
+    - crane tag $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-alpine ${CI_COMMIT_TAG} # /python-gitlab:v1.2.3
+    - crane tag $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-alpine latest           # /python-gitlab:latest
+    - crane tag $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-alpine alpine           # /python-gitlab:alpine
+    - crane tag $CI_REGISTRY_IMAGE:$CI_COMMIT_TAG-slim slim               # /python-gitlab:slim
```
Comment (JohnVillalovos) [inline .gitlab-ci.yml:29]: Probably dumb question. I see how it gets tagged as for example: `python-gitlab:v3.1.0-slim` but how does it get tagged as `python-gitlab:slim` ?

Comment (JohnVillalovos) [reply .gitlab-ci.yml:29]: My assumption is that line 29 here tags it as `python-gitlab:v3.1.0-slim`

Comment (JohnVillalovos) [reply .gitlab-ci.yml:29]: Ah. Line 15 is what does that tag. And line 29 is tagging it as `python-gitlab:slim`

Comment (nejch) [reply .gitlab-ci.yml:29]: So kaniko with a `--destination` flag will automatically push to the registry with what's defined there that's where we tag it with the two `:${CI_COMMIT_TAG}-<flavor>` tags. After the first stage, we have these 2 in the registry.

So by the time we are calling `crane tag`, those version-tagged images are already in the registry and `crane tag` just re-applies these tags in this job remotely (more efficient than doing it locally).

kaniko docs:
https://github.com/GoogleContainerTools/kaniko

crane tag:
https://github.com/google/go-containerregistry/blob/main/cmd/crane/doc/crane_tag.md#crane-tag

File: README.rst [hunk 3]
```diff
@@ -53,8 +53,24 @@ From GitLab:
...
-Using the docker image
-----------------------
+Using the docker images
+-----------------------
+
+``python-gitlab`` provides Docker images in two flavors, based on the ``-alpine`` and ``-slim``
+python `base images <https://hub.docker.com/_/python>`__. The default tag is alpine, but you
+can explicitly use the alias (see below).
+
+The alpine image is smaller, but you may want to use the Debian-based ``-slim`` tag if you are
```
Comment (max-wittig) [inline README.rst:63]: That is confusing

Comment (max-wittig) [reply README.rst:63]: We have a smaller image, which is not called `slim` and the bigger image is called `slim` 🤔

Comment (nejch) [reply README.rst:63]: Yes, I was also pondering this in the issue https://github.com/python-gitlab/python-gitlab/issues/1852#issuecomment-1017393648 :( will try to come up with a better wording

Comment (nejch) [reply README.rst:63]: It's common to have `:alpine` and `:slim` with alpine being smaller, so maybe I can just add that `latest` is the same as `alpine` for backward compatibility?

Comment (nejch) [reply README.rst:63]: @max-wittig or I could simple move `:slim` to `:slim-bullseye` and then it's clear that "this is a slim debian image"?

https://github.com/docker-library/python/blob/d7fa897e3d1c57cf144b6253ffb08a11b7a7511c/3.10/bullseye/slim/Dockerfile

Comment (nejch) [reply README.rst]: @max-wittig switched to `slim-bullseye` to be more explicit about the base image.

Comment (nejch) [reply README.rst]: Looking at this again I don't really like it as people will get stuck on old images then, I'll take another look

Comment (nejch) [reply README.rst]: So now we just follow the python flavor tags, I've documented that. And we can upgrade the debian dist every few years manually.

**Review by JohnVillalovos — APPROVED** [review]

**Pull request merged**